### PR TITLE
send unhealthy endpoints to Envoy (take 2)

### PR DESCRIPTION
**Pull request by ramaraochavali**
- [ ] Configuration Infrastructure
- [ ] Docs
- [ ] Installation
- [X] Networking
- [ ] Performance and Scalability
- [ ] Policies and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure

### CODE REVIEW

File: pilot/pkg/xds/eds.go [hunk 1]
```diff
@@ -122,32 +136,71 @@ func (s *DiscoveryServer) EDSCacheUpdate(shard model.ShardKey, serviceName strin
...
 	ep.Shards[shard] = istioEndpoints
 	// Check if ServiceAccounts have changed. We should do a full push if they have changed.
 	saUpdated := s.UpdateServiceAccount(ep, hostname)
+
+	// For existing endpoints, we need to do full push if service accounts change.
+	if saUpdated {
+		log.Infof("Full push, service accounts changed, %v", hostname)
+		pushType = FullPush
+	}
+	if features.SendUnhealthyEndpoints && pushType == IncrementalPush {
```
Comment (ramaraochavali) [inline pilot/pkg/xds/eds.go:173]: I am suspecting this computation outside of lock might be creating issues - So moved inside lock

**Comment by ramaraochavali** [comment]
@howardjohn made some changes to edsCacheUpdate function - not 100% sure if it is cause. But we can certainly try - WDYT?

File: pilot/pkg/xds/eds.go (hunk 1 above)
Comment (bianpengyuan) [reply pilot/pkg/xds/eds.go:173]: But in the original PR, SA update from 0 to 1 should trigger an early return and a full push, even before this optimization computation right? And the service account computation at cluster is independent of the shard cache here...

Comment (ramaraochavali) [reply pilot/pkg/xds/eds.go:173]: Correct. As long as full push is triggered, the service accounts should be obtained from push context and delivered.

**Comment by ramaraochavali** [comment]
@bianpengyuan @howardjohn I found another issue where instancesByPort which is used to populate service accounts in cluster is not handling unready endpoints - this could be one possibility. PTAL.

File: pilot/pkg/serviceregistry/kube/controller/endpoints.go [hunk 2]
```diff
@@ -208,6 +182,42 @@ func (e *endpointsController) buildIstioEndpoints(endpoint interface{}, host hos
 	return endpoints
 }
 
+func (e *endpointsController) buildServiceInstances(ep *v1.Endpoints, ss v1.EndpointSubset, endpoints []v1.EndpointAddress,
+	svc *model.Service, discoverabilityPolicy model.EndpointDiscoverabilityPolicy, labelsList labels.Collection,
+	svcPort *model.Port, health model.HealthStatus) []*model.ServiceInstance {
```
Comment (bianpengyuan) [inline pilot/pkg/serviceregistry/kube/controller/endpoints.go:187]: I think I must miss something but I don't think `health` is used in this function?

Comment (ramaraochavali) [reply pilot/pkg/serviceregistry/kube/controller/endpoints.go:187]: it is used 
istioEndpoint.HealthStatus = health

Comment (bianpengyuan) [reply pilot/pkg/serviceregistry/kube/controller/endpoints.go:187]: hmm I don't see `health` between line 188 to 219.

Comment (bianpengyuan) [reply pilot/pkg/serviceregistry/kube/controller/endpoints.go:187]: Oh I see, I must look at an old commit.

**Review by bianpengyuan — APPROVED** [review]

File: pilot/pkg/xds/eds.go [hunk 3]
```diff
@@ -161,13 +183,39 @@ func (s *DiscoveryServer) edsCacheUpdate(shard model.ShardKey, hostname string,
...
+		// identify based on health status and push too many times - which is ok since its an optimization.
+		emap := make(map[string]*model.IstioEndpoint, len(oldIstioEndpoints))
+		for _, oie := range oldIstioEndpoints {
+			emap[oie.Address] = oie
+		}
+		needPush := false
+		for _, nie := range istioEndpoints {
+			if oie, exists := emap[nie.Address]; exists {
+				// If endpoint exists already, we should push if it's health status changes.
+				needPush = oie.HealthStatus != nie.HealthStatus
```
Comment (bianpengyuan) [inline pilot/pkg/xds/eds.go:200]: IIUC the original logic will do incremental push if there is any endpoint update event, and not limit to health status change. Now, we will only increment push if health status changes or there is new healthy endpoint. I am not sure if there is other part of EDS resource that could be changed due to endpoint update event?

Comment (ramaraochavali) [reply pilot/pkg/xds/eds.go:200]: That did not change. It is same.

File: pilot/pkg/xds/eds.go [hunk 4]
```diff
@@ -175,48 +183,39 @@ func (s *DiscoveryServer) edsCacheUpdate(shard model.ShardKey, hostname string,
...
+			if needPush {
+				break
+			}
 		}
-	}
 
-	if needPush {
-		return IncrementalPush
+		if !needPush {
+			log.Infof("No push, either old endpoint health status did not change or new endpoint came with unhealthy status, %v", hostname)
```
[the remark this answers is not in the document]
Comment (ramaraochavali) [reply pilot/pkg/xds/eds.go:213]: I will move to debug - just want to validate once with postSubmit.

**Comment by ramaraochavali** [comment]
@howardjohn Do you think this is safe to try in post submit?

**Review by howardjohn — APPROVED** [review]

**Comment by ramaraochavali** [comment]
@howardjohn are you seeing any postsubmit issues after this merge?

**Comment by howardjohn** [comment]
nope!

**Comment by ramaraochavali** [comment]
great. Thank you.

**Comment by ramaraochavali** [comment]
> Seems we are sending traffic to not ready pod. This should never happen

That is strange. Envoy should not send if we send it as Unhealthy. Do you have config_dump of endpoints?

**Comment by howardjohn** [comment]
I don't but it's trivial to reproduce. Just scale from 0 to 1 while sending requests (I did it via ingress).

Maybe healthy panic threshold?

**Comment by ramaraochavali** [comment]
May be. If you have eds config dump - please share?

**Comment by howardjohn** [comment]
```
2022-03-09T10:34:25.495171 outbound|80||echo.default.svc.cluster.local::observability_name::outbound|80||echo.default.svc.cluster.local
2022-03-09T10:34:25.495231 outbound|80||echo.default.svc.cluster.local::default_priority::max_connections::4294967295
2022-03-09T10:34:25.495243 outbound|80||echo.default.svc.cluster.local::default_priority::max_pending_requests::4294967295
2022-03-09T10:34:25.495252 outbound|80||echo.default.svc.cluster.local::default_priority::max_requests::4294967295
2022-03-09T10:34:25.495261 outbound|80||echo.default.svc.cluster.local::default_priority::max_retries::4294967295
2022-03-09T10:34:25.495276 outbound|80||echo.default.svc.cluster.local::high_priority::max_connections::1024
2022-03-09T10:34:25.495286 outbound|80||echo.default.svc.cluster.local::high_priority::max_pending_requests::1024
2022-03-09T10:34:25.495295 outbound|80||echo.default.svc.cluster.local::high_priority::max_requests::1024
2022-03-09T10:34:25.495303 outbound|80||echo.default.svc.cluster.local::high_priority::max_retries::3
2022-03-09T10:34:25.495313 outbound|80||echo.default.svc.cluster.local::added_via_api::true
2022-03-09T10:34:25.495323 outbound|80||echo.default.svc.cluster.local::10.244.0.16:80::cx_active::0
2022-03-09T10:34:25.495333 outbound|80||echo.default.svc.cluster.local::10.244.0.16:80::cx_connect_fail::18
2022-03-09T10:34:25.495344 outbound|80||echo.default.svc.cluster.local::10.244.0.16:80::cx_total::18
2022-03-09T10:34:25.495354 outbound|80||echo.default.svc.cluster.local::10.244.0.16:80::rq_active::0
2022-03-09T10:34:25.495364 outbound|80||echo.default.svc.cluster.local::10.244.0.16:80::rq_error::18
2022-03-09T10:34:25.495374 outbound|80||echo.default.svc.cluster.local::10.244.0.16:80::rq_success::0
2022-03-09T10:34:25.495384 outbound|80||echo.default.svc.cluster.local::10.244.0.16:80::rq_timeout::0
2022-03-09T10:34:25.495394 outbound|80||echo.default.svc.cluster.local::10.244.0.16:80::rq_total::0
2022-03-09T10:34:25.495404 outbound|80||echo.default.svc.cluster.local::10.244.0.16:80::hostname::
2022-03-09T10:34:25.495414 outbound|80||echo.default.svc.cluster.local::10.244.0.16:80::health_flags::/failed_eds_health
2022-03-09T10:34:25.495424 outbound|80||echo.default.svc.cluster.local::10.244.0.16:80::weight::1
2022-03-09T10:34:25.495434 outbound|80||echo.default.svc.cluster.local::10.244.0.16:80::region::
2022-03-09T10:34:25.495444 outbound|80||echo.default.svc.cluster.local::10.244.0.16:80::zone::
2022-03-09T10:34:25.495454 outbound|80||echo.default.svc.cluster.local::10.244.0.16:80::sub_zone::
2022-03-09T10:34:25.495464 outbound|80||echo.default.svc.cluster.local::10.244.0.16:80::canary::false
2022-03-09T10:34:25.495474 outbound|80||echo.default.svc.cluster.local::10.244.0.16:80::priority::0
2022-03-09T10:34:25.495484 outbound|80||echo.default.svc.cluster.local::10.244.0.16:80::success_rate::-1.0
2022-03-09T10:34:25.495494 outbound|80||echo.default.svc.cluster.local::10.244.0.16:80::local_origin_success_rate::-1.0
```

**Comment by howardjohn** [comment]
```
2022-03-09T10:36:58.626834      "endpoint_config": {
2022-03-09T10:36:58.626844       "@type": "type.googleapis.com/envoy.config.endpoint.v3.ClusterLoadAssignment",
2022-03-09T10:36:58.626853       "cluster_name": "outbound|80||echo.default.svc.cluster.local",
2022-03-09T10:36:58.626863       "endpoints": [
2022-03-09T10:36:58.626873        {
2022-03-09T10:36:58.626882         "locality": {},
2022-03-09T10:36:58.626892         "lb_endpoints": [
2022-03-09T10:36:58.626903          {
2022-03-09T10:36:58.626913           "endpoint": {
2022-03-09T10:36:58.626923            "address": {
2022-03-09T10:36:58.626933             "socket_address": {
2022-03-09T10:36:58.626943              "address": "10.244.0.18",
2022-03-09T10:36:58.626952              "port_value": 80
2022-03-09T10:36:58.626962             }
2022-03-09T10:36:58.626972            },
2022-03-09T10:36:58.626982            "health_check_config": {}
2022-03-09T10:36:58.626991           },
2022-03-09T10:36:58.627001           "health_status": "UNHEALTHY",
2022-03-09T10:36:58.627011           "metadata": {
2022-03-09T10:36:58.627021            "filter_metadata": {
2022-03-09T10:36:58.627031             "istio": {
2022-03-09T10:36:58.627040              "workload": "echo;default;echo;latest;Kubernetes"
2022-03-09T10:36:58.627050             },
2022-03-09T10:36:58.627060             "envoy.transport_socket_match": {
2022-03-09T10:36:58.627070              "tlsMode": "istio"
2022-03-09T10:36:58.627080             }
2022-03-09T10:36:58.627089            }
2022-03-09T10:36:58.627099           },

```

same thing in different form

**Comment by ramaraochavali** [comment]
So this seems correct from our side. May be healthy panic as you suspected? Let me spent time on it. Not sure what our options are here if it is that.

**Comment by howardjohn** [comment]
one thing is I thought we did not send the initial unready update. And only
send it if we transition from ready to not ready. Did we change that?

**Comment by ramaraochavali** [comment]
> one thing is I thought we did not send the initial unready update. And only
send it if we transition from ready to not ready. Did we change that?

I remember that too. Let me take a look at the code

**Comment by ramaraochavali** [comment]
https://github.com/istio/istio/pull/37854 - @howardjohn PtAL

**Pull request merged**